### DISCUSSION TITLE
Adding custom column to _users collection and updating the column value of a certain user but I can't pull anything changed!

### DISCUSSION THREAD

**Discussion by cansati** [pocketbase/pocketbase#956] [Q&A] [answered]
Hello everyone, I've just started to discover amazing pocketbase as a go framework. I'm trying to add a custom column ("role") into _users collection and with the help of OnUserAfterCreateRequest api event triggered hook, I'm trying to change the "role" column value of the user after user creation process is completed. I'm using the code below to add the column:
`app.Dao().DB().AddColumn("_users", "role", "text").Execute()`

When I checked it with the code below:
`cols, err = app.Dao().GetTableColumns("_users")`
`log.Println(cols)` prints ->[id verified email tokenKey passwordHash lastResetSentAt lastVerificationSentAt created updated role]
The "role" column is at the end of the slice. OK.

Afterwards I'm updating the "role" column of the user with the code below:
`_, err = app.Dao().DB().Update("_users", dbx.Params{"role": "consumer"}, dbx.HashExp{"id": e.User.Id}).Execute()`

When I checked the user with the code below:
```
user := models.User{}
err = app.Dao().DB().
	NewQuery("SELECT * FROM _users WHERE id={:id}").
	Bind(dbx.Params{"id": e.User.Id}).
	One(&user)
```
`log.Println(user.AsMap())`
-> map[created:2022-11-07 10:30:35.835 email:asdf1@gmail.com id:l1vlhzstuvgoo0c lastResetSentAt: lastVerificationSentAt: profile:<nil> updated:2022-11-07 10:30:35.835 verified:false] <nil>

There is nothing about "role" column associated with the user. What am I doing wrong? Thanks for any help.

**Comment by ganigeorgiev** [comment, marked as answer]
@cansati I don't recommend manually changing the internal tables (those prefixed with `_`). I guess you are using PocketBase v0.7.x. The `User` model and the related `_users` table is fixed and you can't just add dynamic columns to it.

If you want to add a `role` column/field, the recommended way would be to define it as part of the `profiles` collection related to your users. You can do that from the "Admin UI -> Users -> Edit profiles collection -> Add new field". If you want to do it programmatically you'll have to update the `profiles` collection from the Dao:

```go
// fetch the profiles collection
profiles, err := app.Dao().FindCollectionByNameOrId("profiles")

// register new field
profiles.Schema.AddField(&schema.SchemaField{
    Type: schema.FieldTypeText,
    Name: "role",
    // etc.
})

// save the updated collection
if err := app.Dao().SaveCollection(profiles); err != nil {
    return err
}
```

The `role` would be part of the `profiles` collection. Then if you want to overwrite the submitted user role, it is better to use the before hook, aka, `OnRecordBeforeUpdateRequest` since it runs right after request data load and before persisting the model. The after hook executes after the model is persisted in the database and you can't change the returned response to the client from there in case of an error.

```go
app.OnRecordBeforeUpdateRequest().Add(func(e *core.RecordUpdateEvent) error {
	// note that there is no need to explicitly persist the value
	// because the finalizer of the before hook will do it for us if
	// all attached events retun no error
	e.Record.SetDataValue("role", "consumers") // in v0.8.0 this method is just Set

	return nil
})
```

---

**Please note that the above has changed in v0.8.0-rc and the `users` and `profiles` are now merged per #376. You can read more about the changes in https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc1**

**Reply by cansati** [reply]
@ganigeorgiev I updated my pocketbase to v0.8.0-rc2.
When should I add the field programmatically? The next command is not working before any user registration (I think the "users" collection is not initializing completely before registering any user). And I can't register any user with "role" value without having the "role" field defined at the users collection.
 `users, err := app.Dao().FindCollectionByNameOrId("users")`

Also I don't want to execute the `e.Record.Set("role", "consumer")` every time when any record created but OnRecordBeforeCreateRequest is not giving me any filter choice like "when creation of user record". Whan can i do about it? I believe there is a way I couldn't see.

**Reply by ganigeorgiev** [reply]
> When should I add the field programmatically? The next command is not working before any user registration (I think the "users" collection is not initializing completely before registering any user). And I can't register any user with "role" value without having the "role" field defined at the users collection.

I'm confused. When making db/schema changes they should be done only once, not on every request. This means that your new field need to be registered either via the "Admin UI -> Collections -> Users -> Edit collection" or in a [DB migration file](https://pocketbase.io/docs/db-migrations/). My recommendtion would be to create the field from the UI and then run `migrate collection` to create an automated migration file with your latest collections changes. More info could be found in https://pocketbase.io/docs/db-migrations/#collections-migration. There will be more examples on this after the final v0.8.0 release.

---

> Also I don't want to execute the e.Record.Set("role", "consumer") every time when any record created but OnRecordBeforeCreateRequest is not giving me any filter choice like "when creation of user record". Whan can i do about it? I believe there is a way I couldn't see.

You can check the record collection name and further filter the hook event:

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
    if e.Record.Collection().Name != "users" {
        return nil // not an user collection
    }

    // note that there is no need to explicitly persist the value
    // because the finalizer of the before hook will do it for us if
    // all attached events retun no error
    e.Record.Set("role", "consumers")

    return nil
})
```

**Reply by cansati** [reply]
You've clarified everything.

Thank you @ganigeorgiev you are awesome 🙃
